kotlinx-coroutines-core/common/src/internal/SegmentList.kt

```diff
@@ -0,0 +1,213 @@
+    while (true) {
+        val s = startFrom.findSegmentInternal(id, createNewSegment)
+        if (s.isClosed || moveForward(s.segment)) return s
+    }
+}
+
+/**
+ * Closes this linked list of segments by forbidding adding new segments,
+ * returns the last segment in the list.
+ */
+internal fun <S : Segment<S>> S.close(): S {
+    var cur: S = this
+    while (true) {
+        val next = cur.nextOrClosed.run { if (isClosed) return cur else segment }
+        if (next === null) {
+            if (cur.markAsClosed()) return cur
+        } else {
+            cur = next
+        }
+    }
+}
+
+/**
+ * Each segment in [SegmentList] has a unique id and is created by [SegmentList.newSegment].
+ * Essentially, this is a node in the Michael-Scott queue algorithm, but with
+ * maintaining [prev] pointer for efficient [remove] implementation.
+ */
+internal abstract class Segment<S : Segment<S>>(val id: Long, prev: S?, pointers: Int) {
+    // Pointer to the next segment, updates similarly to the Michael-Scott queue algorithm.
+    private val _next = atomic<Any?>(null)
+    val nextOrClosed: NextSegmentOrClosed<S> get() = NextSegmentOrClosed(_next.value)
+    fun trySetNext(value: S): Boolean = _next.compareAndSet(null, value)
+
+    // Pointer to the previous segment, updates in [remove] function.
+    val prev = atomic(prev)
+
+    /**
+     * Cleans the pointer to the previous segment.
+     */
+    fun cleanPrev() = prev.lazySet(null)
+
+    /**
+     * This property should return the maximal number of slots in this segment,
+     * it is used to define whether the segment is logically removed.
+     */
+    abstract val maxSlots: Int
+
+    // numbers of cleaned slots (lowest bits) and AtomicRef pointers to this segment (highest bits)
+    private val cleanedAndPointers = atomic(pointers shl POINTERS_SHIFT)
+
+    /**
+     * Returns `true` if this segment is logically removed from the queue.
+     * The segment is considered as removed if all the slots are cleaned,
+     * there is no pointers to this segment from outside, and
+     * it is not a physical tail in the linked list of segments.
+     */
+    inline val removed get() = cleanedAndPointers.value == maxSlots && _next.value !== null
+
+    // increments the number of pointers if this segment is not logically removed\
+    inline fun tryIncPointers() = cleanedAndPointers.addConditionally(1 shl POINTERS_SHIFT) { it != maxSlots }
+
+    // returns `true` if this segment is logically removed after the decrement
+    inline fun decPointers() = cleanedAndPointers.addAndGet(-(1 shl POINTERS_SHIFT)) == maxSlots
+
+    /**
+     * This functions should be invoked on each slot clean-up;
+     * should not be invoked twice for the same slot.
+     */
+    fun onSlotCleaned() {
+        if (cleanedAndPointers.incrementAndGet() == maxSlots) remove()
```

Comment:
In the same vein, this method can be called on a physical tail, and `remove()` may be run even though `removed` is not yet true. Hence, the assertion in `remove()` may fail. Once again, it is fine if we make certain assumptions about how pointers and cells behave: that is, the pointers only move to the right and never disappear, and no cells in segment `n` may be cancelled unless some pointer points to a segment with id of `n` or above. Still, this should probably be kept in mind.

Reply:
See https://github.com/Kotlin/kotlinx.coroutines/pull/1563#discussion_r345505399